### PR TITLE
Support stack-allocating functions when allocating heap vars

### PR DESCRIPTION
**Pull request by mrstanb**
During the SV-COMP benchmark runs, we noticed that calls to `alloca()` and `__builtin_alloca()` result in heap variables being allocated during analysis. However, when passing memory, allocated with these function, to `free()` no warning was produced, since Goblint couldn't differentiate between stack- and heap-allocated heap vars.

This PR:
* Updates `Queries.HeapVar`, s.t. it now takes a boolean flag indicating whether the heap variable in the query answer is allocated on the stack or the heap
* Adds `Queries.IsDynamicallyAlloced` which works like `Queries.IsHeapVar` with the difference that it also differentiates whether the heap var is one abstracting stack or heap memory

TODOs:
* [x] Track memory, allocated via `alloca()`-like functions, for each function that calls `alloca()`-like functions
* [x] Make sure that `useAfterFree` can free memory, allocated via `alloca()`-like functions, as soon as a function that called `alloca()` is returned from

### CODE REVIEW

File: src/analyses/wrapperFunctionAnalysis.ml [hunk 1]
```diff
@@ -133,17 +133,20 @@ module MallocWrapper : MCPSpec = struct
...
       in
       let count = UniqueCallCounter.find (`Lifted node) counter in
       let var = NodeVarinfoMap.to_varinfo (ctx.ask Q.CurrentThreadId, node, count) in
       var.vdecl <- UpdateCil.getLoc node; (* TODO: does this do anything bad for incremental? *)
+      if on_stack then var.vattr <- addAttribute (Attr ("stack_alloca", [])) var.vattr; (* If the call was for stack allocation, add an attr to mark the heap var *)
       `Lifted var
     | Q.IsHeapVar v ->
       NodeVarinfoMap.mem_varinfo v
+    | Q.IsDynamicallyAlloced v ->
+      NodeVarinfoMap.mem_varinfo v && not @@ hasAttribute "stack_alloca" v.vattr
```
Comment (michael-schwarz) [inline src/analyses/wrapperFunctionAnalysis.ml:149]: In what sense are stack things allocated by `alloca` not dynamically allocated? I think the terminology is a bit confusing here.

File: src/analyses/libraryFunctions.ml [hunk 2]
```diff
@@ -64,6 +64,7 @@ let c_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
     ("__builtin_strcmp", special [__ "s1" [r]; __ "s2" [r]] @@ fun s1 s2 -> Strcmp { s1; s2; n = None; });
     ("strncmp", special [__ "s1" [r]; __ "s2" [r]; __ "n" []] @@ fun s1 s2 n -> Strcmp { s1; s2; n = Some n; });
     ("malloc", special [__ "size" []] @@ fun size -> Malloc size);
+    ("alloca", special [__ "size" []] @@ fun size -> Malloc size); (* TODO: Maybe define a new special type [Alloca], just like [Malloc]? *)
```
Comment (michael-schwarz) [inline src/analyses/libraryFunctions.ml:67]: Yes, having a special type here makes sense!

File: src/analyses/wrapperFunctionAnalysis.ml (hunk 1 above)
Comment (mrstanb) [reply src/analyses/wrapperFunctionAnalysis.ml:149]: You're right, `alloca` is indeed dynamic allocation as well. I got `HeapVar` and `IsDynamicallyAlloced` the other way around.

Comment (mrstanb) [reply src/analyses/wrapperFunctionAnalysis.ml]: `IsHeapVar` and `IsDynamicallyAlloced` should be now implemented the other way around. All of their usages in Goblint should be also adapted to these changes now. Cf. https://github.com/goblint/analyzer/pull/1179/commits/d9df607e99ea78600b81f879955208ec00391bcb

File: src/analyses/libraryFunctions.ml (hunk 2 above)
Comment (mrstanb) [reply src/analyses/libraryFunctions.ml]: Added `Alloca` type in https://github.com/goblint/analyzer/pull/1179/commits/aa97af86c23911bd05822149a7d9c9fc4f819ae2. I also moved `__builtin_alloca`, s.t. it's now classified only as `Alloca` and not as `Malloc` anymore: https://github.com/goblint/analyzer/pull/1179/commits/6e9bb2f121114a9c91bf1bb19557bbcfd0d3ffca

**Comment by mrstanb** [comment]
I'm only wondering about one thing, namely `__builtin_alloca` is now defined
* Once as part of `c_descs_list`
* And once as part of `gcc_descs_list`

Iirc, `__builtin_*` functions are GCC-specific and thus it's probably worth it to keep `__builtin_alloca` only in `gcc_descs_list`. What do you think?

**Comment by sim642** [comment]
Indeed there should be just one definition of `__builtin_alloca` and it should be under GCC. Also `alloca` itself should go there because it's not part of any C standard. According to the man page, it's not part of any standard, but is "machine- and compiler-dependent".

File: src/analyses/base.ml [hunk 3]
```diff
@@ -150,8 +150,8 @@ struct
 
   let longjmp_return = ref dummyFunDec.svar
 
-  let heap_var ctx =
-    let info = match (ctx.ask Q.HeapVar) with
+  let heap_var on_stack ctx =
+    let info = match (ctx.ask (Q.HeapVar {on_stack = on_stack})) with
```
Comment (sim642) [inline src/analyses/base.ml:154]:
```suggestion
    let info = match (ctx.ask (Q.HeapVar {on_stack})) with
```

File: src/analyses/base.ml [hunk 4]
```diff
@@ -1249,7 +1249,7 @@ struct
           (* If there's a non-heap var or an offset in the lval set, we answer with bottom *)
           (* If we're asking for the BlobSize from the base address, then don't check for offsets => we want to avoid getting bot *)
           if AD.exists (function
-              | Addr (v,o) -> (not @@ ctx.ask (Queries.IsHeapVar v)) || (if not from_base_addr then o <> `NoOffset else false)
+              | Addr (v,o) -> (not @@ ctx.ask (Queries.IsDynamicallyAlloced v)) || (ctx.ask (Queries.IsDynamicallyAlloced v) && not @@ ctx.ask (Queries.IsHeapVar v)) || (if not from_base_addr then o <> `NoOffset else false)
```
Comment (sim642) [inline src/analyses/base.ml:1252]: Should use parenthesis instead of `@@`s here because it's not obvious where the `@@` ends.
Also, the query is duplicated, so it would be nice to just use a variable for it and make this line more readable.

**Comment by mrstanb** [comment]
> Indeed there should be just one definition of `__builtin_alloca` and it should be under GCC. Also `alloca` itself should go there because it's not part of any C standard. According to the man page, it's not part of any standard, but is "machine- and compiler-dependent".

Both `alloca` and `__builtin_alloca` should be now part of `gcc_descs_list` only. Cf. https://github.com/goblint/analyzer/pull/1179/commits/35513d09c2ed2a0fb6f00a1e0dd1854f9dd79329

File: src/domains/queries.ml [hunk 5]
```diff
@@ -101,8 +101,9 @@ type _ t =
   | IterVars: itervar -> Unit.t t
   | PathQuery: int * 'a t -> 'a t (** Query only one path under witness lifter. *)
   | DYojson: FlatYojson.t t (** Get local state Yojson of one path under [PathQuery]. *)
-  | HeapVar: VI.t t
+  | HeapVar: {on_stack: bool} -> VI.t t (* If on_stack is [true], then alloca() or a similar function was called *)
   | IsHeapVar: varinfo -> MayBool.t t (* TODO: is may or must? *)
+  | IsDynamicallyAlloced: varinfo -> MayBool.t t (* [true] if heap var represents dynamically alloced memory *)
```
Comment (sim642) [inline src/domains/queries.ml:106]: The naming is very confusing. If I understand correctly, `alloca` also uses `HeapVar`. But `alloca` doesn't allocate on the heap, it allocates on the stack!
And the difference between `IsHeapVar` and `IsDynamicallyAlloced` is completely unclear to me. Both `malloc` and `alloca` variables are created by `HeapVar`, so it seems like `IsHeapVar` is true for both. But both are also dynamically allocated.

Since these "heap" variables get `(alloc@...)` as their name, maybe their creation query should be `AllocVar` and the query corresponding to it `IsAllocVar`. Then `IsHeapVar` would distinguish those by `malloc` from those by `alloca`, or something.

Comment (mrstanb) [reply src/domains/queries.ml:106]: I agree that the naming might be confusing. `HeapVar` should be now referring to any dynamic allocation-site abstraction (i.e., `(alloca@...)`), which can in particular be `alloca` (on the stack) and `malloc` (on the heap).
Hence, I agree with the `AllocVar` name in place of `HeapVar`.

`IsDynamicallyAlloced` essentially checks if a var refers to a dynamic allocation-site abstraction, hence we could change it to `IsAllocVar`. `IsHeapVar` should be already be able to distinguish between `alloc@...` vars from `alloca` and from `malloc`. Iirc, I added an extra `vattr` to the `alloc@...` vars for this which is implemented in `HeapVar`.

In short, my suggestion:
* Rename `HeapVar` -> `AllocVar`
* Rename `IsDynamicallyAlloced` -> `IsAllocVar`
* Keep `IsHeapVar` as is
* (not sure if it'd be needed, but we could also add `IsStackVar` which is dual to `IsHeapVar`)

Please let me know, what do you think. CC @michael-schwarz, because we also discussed the naming with him before.

Comment (michael-schwarz) [reply src/domains/queries.ml:106]: Yes, seems reasonable!

File: src/analyses/base.ml (hunk 4 above)
Comment (mrstanb) [reply src/analyses/base.ml]: Should be fixed in https://github.com/goblint/analyzer/pull/1179/commits/3bdc92d7c6e58520f6ed552cb687fa158adcf429

File: src/domains/queries.ml (hunk 5 above)
Comment (mrstanb) [reply src/domains/queries.ml]: I did the renaming in https://github.com/goblint/analyzer/pull/1179/commits/7b1d29850036fb1287b10c8aeb962a162b7fac11

**Review by michael-schwarz — APPROVED** [review]

**Comment by michael-schwarz** [comment]
Sorry for not waiting on your review @sim642, but things are starting to get urgent in order for @mrstanb's changes being all set and integrated with the rest in time for us to still run benchmarks without opening another staging branch and the hassle that creates.
If it turns out later that there's anything that should still be fixed here Stanimir (or I) will address it on master / in a follow-up PR.
